sdks/python/apache_beam/testing/util.py

```diff
@@ -120,3 +125,66 @@ def open_shards(glob_pattern):
       f.write(file(shard).read())
     concatenated_file_name = f.name
   return file(concatenated_file_name, 'rb')
+
+
+class _WindowedValueMatcherDoFn(DoFn):
+  """Verifies that all processed elements are in the list of expected
+  elements."""
+
+  def __init__(self, expected_elements):
+    """
+    Arguments:
+      expected_elements: A list of (element, timestamp, window) tuples to match.
+    """
+    super(_WindowedValueMatcherDoFn, self).__init__()
+    self.expected_elements = copy.copy(expected_elements)
+
+  def process(self, element, timestamp=DoFn.TimestampParam,
```

Comment:
This will not work if there are multiple bundles. 

I would propose that WindowedValueMatcher instead be implemented on top of assert_that, where the windows and timestamps are reified. Similar to how reify_windows is used in https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/window_test.py . If we want to make this a full util, we should probably provide a similar API to assert_that.

Reply:
I believe this code does work with multiple bundles. It "Verifies that all processed elements are in the list of expected
elements."

That being said, I could try to rewrite this to use assert_that.

Reply:
I should clarify, it's use in WindowedValueMatcher is not correct in the presence of multiple bundles. E.g. suppose the expected values were [a, b] and the actual values were [a, a] split into two bundles.